### ISSUE TITLE
Appium taking too long to find an element on a particular screen

### ISSUE THREAD

**Issue by shoaibmansoor** [appium/appium#15366]
## The problem

Appium takes around 125 seconds to find any element by any identifier on a specific page.

## Environment

* Appium version (or git revision) that exhibits the issue: 1.18.0, 1.19.1, and 1.20.2
* Last Appium version that did not exhibit the issue (if applicable): All versions I could test have this issue
* Desktop OS/version used to run Appium: Windows 10
* Node.js version (unless using Appium.app|exe):  Not sure. I am using browserstack
* Npm or Yarn package manager: Not sure. I am using browserstack
* Mobile platform/version under test: iOS
* Real device or emulator/simulator: Real device - iPhone 11(iOS v14.0)
* Appium CLI or Appium.app|exe: Appium CLI

## Details
Appium is taking around 125 seconds to find any element by any identifier on a specific page.

![image](https://user-images.githubusercontent.com/22621280/117970029-ce49f200-b345-11eb-9642-9f9531864d7e.png)

On the same screen, request for page_source is always getting timeout with exception: 
`Appium error: An unknown server-side error occurred while processing the command. Original error: Could not proxy command to the remote server. Original error: timeout of 240000ms exceeded`
Getting a screenshot however works but it also takes around 20 seconds to complete a request.

I have tried several combinations of settings like `{"maxDepth":65,"snapshotMaxDepth":65,"customSnapshotTimeout":0,"snapshotTimeout":0,"useFirstMatch":true}`
by adding/removing and increasing/decreasing the values of these properties, it helps in finding elements at the cost of increased time(125 seconds).
The issue of getting_page source that screen is always there.

## Link to Appium logs

Appium logs for a long time in finding element: 
[appium-server.log](https://github.com/appium/appium/files/6465801/appium-server.log)

Device logs for a long time in finding element: 
[device-log.log](https://github.com/appium/appium/files/6465802/device-log.log)

Appium logs for page_source issue on the same page: 
[appium-server-timeout.log](https://github.com/appium/appium/files/6465804/appium-server-timeout.log)

Device logs for page_source issue on the same page: 
[device-timeout.log](https://github.com/appium/appium/files/6465810/device-timeout.log)

Kindly look into this issue and provide some guidance to resolve it.

Thanks

**Comment by shoaibmansoor** [comment]
With `{"snapshotTimeout":100,"useFirstMatch":true}` settings, I see below message from XCUITest
![image](https://user-images.githubusercontent.com/22621280/117971680-d30fa580-b347-11eb-809e-03624bff0ab0.png)

**Comment by mykola-mokhnach** [comment]
I don't think there's much we could do on Appium side about it, since the elements lookup stuff is controlled by XCTest.
As a workaround you could try to:

- Use the most recent Appium version
- Replace xpath locators with other locator types
- Make sure the application under test does not hog the main event loop constantly while this particular view is visible
- Reduce the amount of elements on this particular view

**Comment by shoaibmansoor** [comment]
@mykola-mokhnach Thanks for the reply. Let me try these options and update you.

**Comment by shoaibmansoor** [comment]
@mykola-mokhnach I tried your suggestions but it looks like I'm out of luck.

> Use the most recent Appium version

I need to execute my tests on browserstack, and the latest appium version supported by them is 1.20.2. Although, I tried locally but the issue is reproducible on v1.21.0 as well.

> Replace xpath locators with other locator types

Unfortunately, Appium gets stuck as soon as that screen appears so no way of getting attributes to locate an element

I am left with the last two options then.

Thanks for your help.

Feel free to close the issue.
